src/transformers/trainer_tf.py

```diff
@@ -384,141 +415,144 @@ def train(self) -> None:
+
+                    if (
+                        self.global_step > 0
+                        and self.args.evaluate_during_training
+                        and self.global_step % self.args.eval_steps == 0
+                    ):
+                        self.evaluate()
+
+                    if (self.global_step > 0 and self.global_step % self.args.logging_steps == 0) or (
+                        self.global_step == 1 and self.args.logging_first_step
+                    ):
+                        logs = {}
+                        logs["loss"] = training_loss.numpy()
+                        logs["learning_rate"] = self.optimizers[1](self.global_step).numpy()
+                        logs["epoch"] = self.epoch_logging
+
+                        self._log(logs)
+
+                    if self.global_step > 0 and self.global_step % self.args.save_steps == 0:
+                        ckpt_save_path = self.model.ckpt_manager.save()
+
+                        logger.info("Saving checkpoint for step {} at {}".format(self.global_step, ckpt_save_path))
+
+                    if self.global_step > 0 and self.global_step % self.steps_per_epoch == 0:
+                        break
+
+                self.train_loss.reset_states()
+
+            end_time = datetime.datetime.now()
+
+            logger.info("Training took: {}".format(str(end_time - start_time)))
 
         if self.args.past_index and hasattr(self, "_past"):
             # Clean the state at the end of training
             delattr(self, "_past")
 
-    def _training_steps(self, ds, optimizer):
-        """
-        Returns a generator over training steps (i.e. parameters update).
-        """
-        for i, loss in enumerate(self._accumulate_next_gradients(ds)):
-            if i % self.args.gradient_accumulation_steps == 0:
-                self._apply_gradients(optimizer)
-                yield loss
-
-    @tf.function
-    def _apply_gradients(self, optimizer):
-        """Applies the gradients (cross-replica)."""
-        self.args.strategy.experimental_run_v2(self._step, args=(optimizer,))
-
-    def _step(self, optimizer):
-        """Applies gradients and resets accumulation."""
-        gradient_scale = self.gradient_accumulator.step * self.args.strategy.num_replicas_in_sync
-        gradients = [
-            gradient / tf.cast(gradient_scale, gradient.dtype) for gradient in self.gradient_accumulator.gradients
-        ]
-        gradients = [(tf.clip_by_value(grad, -self.args.max_grad_norm, self.args.max_grad_norm)) for grad in gradients]
-
-        optimizer.apply_gradients(list(zip(gradients, self.model.trainable_variables)))
-        self.gradient_accumulator.reset()
+    def _create_training_routines(self):
+        with self.args.strategy.scope():
 
-    def _accumulate_next_gradients(self, ds):
-        """Accumulates the gradients from the next element in dataset."""
-        iterator = iter(ds)
+            def training_step(features, labels):
```

Comment:
This part would need to be more easily accessible for subclassing. I don't think it's the case if it's nested in another method.

Reply:
For me this function should not be accessible for subclassing and more precisely should not be changed. Globally, that's why important functions are nested, to do not be updated accidentally.

Reply:
You then remove the ability for any user to add any customization to the way the loss is computed (like adding a penalty to the loss), or should it be done in `run_model`? 

I haven't used TF a lot, but I thought overriding the `train_step` function was the way to go in Keras, so it would be more natural to users to have this name as a customization point.

Reply:
Loss computation is done directly inside the model such as PT ones, you cannot customize it. If you want to modify it, you have to redefine the `run_model` function and update it accordingly to what you want.

Here we are not using the `train_step` method that has been integrated since TF 2.2.

Reply:
Yes loss computation is done directly in the model. I'm talking of the times the user does not want that exactly and wants to customize this. I know we are not using the `train_step` method, just saying a user that discovers our API might look for something with the same name as a customization point instead of `run_model`.

Also the PyTorch side has a `training_step` and a `prediction_step` as points of customization, it would be great if they had the same name for consistency.

Reply:
But the consistency can also be achieved by moving some parts of the PT Trainer API to look more like TF, so there is no obligation to change things here if it does not make any sense.